src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,476 @@
+        }
+
+        protected void WriteType(JsonWriter writer, string key, Type? type)
+        {
+            writer.WriteKey(key);
+            WriteType(writer, type);
+        }
+
+        protected void WriteType(JsonWriter writer, Type? type)
+        {
+            if (type is null)
+            {
+                writer.WriteNull();
+                return;
+            }
+
+            writer.WriteObjectStart();
+            writer.Write("fullName", type.FullName);
+            // Note that the file path to the assembly is deliberately not included here. The file path
+            // of the assembly does not contribute to the output of the program.
+            writer.Write("assemblyName", type.Assembly.FullName);
+            writer.WriteObjectEnd();
+        }
+
+        private (JsonWriter, PooledStringBuilder) CreateWriter()
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            var writer = new StringWriter(builder);
+            return (new JsonWriter(writer), builder);
+        }
+
+        internal string GetKey(
+            Compilation compilation,
+            ImmutableArray<AdditionalText> additionalTexts = default,
+            ImmutableArray<DiagnosticAnalyzer> analyzers = default,
+            ImmutableArray<ISourceGenerator> generators = default,
+            EmitOptions? emitOptions = null,
+            DeterministicKeyOptions options = default)
+        {
+            ensureNotDefault(ref additionalTexts);
+            ensureNotDefault(ref analyzers);
+            ensureNotDefault(ref generators);
+
+            var (writer, builder) = CreateWriter();
+
+            writer.WriteObjectStart();
+
+            writer.WriteKey("compilation");
+            WriteCompilation(writer, compilation, options);
+            writer.WriteKey("additionalTexts");
+            writeAdditionalTexts();
+            writer.WriteKey("analyzers");
+            writeAnalyzers();
+            writer.WriteKey("generators");
+            writeGenerators();
+            writer.WriteKey("emitOptions");
+            WriteEmitOptions(writer, emitOptions);
+
+            writer.WriteObjectEnd();
+
+            return builder.ToStringAndFree();
+
+            void writeAdditionalTexts()
+            {
+                writer.WriteArrayStart();
+                foreach (var additionalText in additionalTexts)
+                {
+                    writer.WriteObjectStart();
+                    WriteFileName(writer, "fileName", additionalText.Path, options);
+                    writer.WriteKey("text");
+                    WriteSourceText(writer, additionalText.GetText());
+                    writer.WriteObjectEnd();
+                }
+                writer.WriteArrayEnd();
+            }
+
+            void writeAnalyzers()
+            {
+                writer.WriteArrayStart();
+                foreach (var analyzer in analyzers)
+                {
+                    WriteType(writer, analyzer.GetType());
```

Comment:
Why do we need to include the names of individual analyzers and generators rather than the list of analyzer/generator assemblies? Also, should we include MVIDs of the assemblies?

Reply:
I think the assemblies make sense from a command line compilation perspective but this can also be used at an API level. It is possible to provide individual analyzers / generators at this level that are not the full set in the containing assembly. Hence using the assembly alone isn't enough. 

Will add the MVID though to `WriteType`